### PR TITLE
Fall back to `pickle` in `protocol.dumps`  instead of failing hard

### LINKED ISSUE #8446: `distributed.protocol.dumps` does not handle objects not serializable via `msgpack.dumps` as suggested by docstring
When using `distributed.protocol.dumps` to serialize arbitrary data, it fails and raises a `TypeError`. The docstring suggests it should be able to handle this.

**Minimal Complete Verifiable Example**:

```python
import numpy as np
from distributed.protocol import dumps

dumps(np.int64(123))
```

raises

```python
2024-01-09 19:55:04,572 - distributed.protocol.core - CRITICAL - Failed to Serialize
Traceback (most recent call last):
  File "/Users/hendrikmakait/projects/dask/distributed/distributed/protocol/core.py", line 108, in dumps
    frames[0] = msgpack.dumps(msg, default=_encode_default, use_bin_type=True)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/mambaforge/base/envs/dask-distributed/lib/python3.12/site-packages/msgpack/__init__.py", line 36, in packb
    return Packer(**kwargs).pack(o)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "msgpack/_packer.pyx", line 294, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 300, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 297, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 291, in msgpack._cmsgpack.Packer._pack
TypeError: can not serialize 'numpy.int64' object
Traceback (most recent call last):
  File "/Users/hendrikmakait/projects/dask/distributed/sandbox.py", line 5, in <module>
    dumps(np.int64(123))
  File "/Users/hendrikmakait/projects/dask/distributed/distributed/protocol/core.py", line 108, in dumps
    frames[0] = msgpack.dumps(msg, default=_encode_default, use_bin_type=True)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/mambaforge/base/envs/dask-distributed/lib/python3.12/site-packages/msgpack/__init__.py", line 36, in packb
    return Packer(**kwargs).pack(o)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "msgpack/_packer.pyx", line 294, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 300, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 297, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 291, in msgpack._cmsgpack.Packer._pack
TypeError: can not serialize 'numpy.int64' object
```

Relevant section of `dumps`'s docstring: https://github.com/dask/distributed/blob/7562f9c566978de4f3f5b73920a24ea1813d6e28/distributed/protocol/core.py#L30-L35

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #8446

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/protocol/core.py [hunk 1]
```diff
@@ -104,7 +105,36 @@ def _encode_default(obj):
             else:
                 return msgpack_encode_default(obj)
 
-        frames[0] = msgpack.dumps(msg, default=_encode_default, use_bin_type=True)
+        try:
+            frames[0] = msgpack.dumps(msg, default=_encode_default, use_bin_type=True)
+        except TypeError as e:
+            logger.warning(e)
```
Comment (fjetter) [inline distributed/protocol/core.py:111]: I'm not sure how meaningful this is to users. Just logging the original exception appears to be an error for users but we're handling it gracefully. If anything we should log a custom message making them aware of a possible performance issue but I'm not sure if this is actually worth a warning log

File: distributed/protocol/core.py [hunk 2]
```diff
@@ -104,7 +105,36 @@ def _encode_default(obj):
...
+                    frames.extend(create_pickled_sub_frames(obj))
+                    return {"__Pickled__": offset}
+                else:
+                    encoded = msgpack_encode_default(obj)
+                    if encoded is not obj or _is_msgpack_serializable(obj):
+                        return encoded
+                    obj = ToPickle(obj)
+                    offset = len(frames)
+                    frames.extend(create_pickled_sub_frames(obj))
+                    return {"__Pickled__": offset}
```
Comment (fjetter) [inline distributed/protocol/core.py:129]: in the interest of DRY

```suggestion
                rv = _enode_default(obj)
                if rv is obj and not _is_msgpack_serializable(obj):
                    obj = ToPickle(obj)
                    offset = len(frames)
                    frames.extend(create_pickled_sub_frames(obj))
                    return {"__Pickled__": offset}
                return rv
```

File: distributed/protocol/core.py (hunk 1 above)
Comment (hendrikmakait) [reply distributed/protocol/core.py:111]: Fair enough, I'll de-escalate that to an INFO log with a custom message. I'd like to have something logged here to raise user attention and possibly for log analysis to highlight potential performance issues.

File: distributed/protocol/core.py (hunk 2 above)
Comment (hendrikmakait) [reply distributed/protocol/core.py:129]: Good point

**Review by fjetter — APPROVED** [review]

**Pull request merged**